### PR TITLE
Add cdemu packages and module

### PR DESCRIPTION
**Pull request by bendlas**
This adds cdemu to nixos. Everything works for me, but it is my first contribution, so there might be some stylistic issues. Especially the desktop integration is a bit fishy. Maybe somebody with experience in packaging gnome apps will know a better way to set up the gtk3 gsettings schemas for the image analyzer.

### CODE REVIEW

File: nixos/modules/services/hardware/cdemu.nix [hunk 1]
```diff
@@ -0,0 +1,49 @@
+{ config, lib, pkgs, ... }:
```
Comment (lethalman) [inline nixos/modules/services/hardware/cdemu.nix]: I would call this file under `programs/cdemu.nix`. It's not a service, it's a series of config to make a program work.

File: nixos/modules/services/hardware/cdemu.nix [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+with lib;
+
+let cfg = config.services.cdemu;
+in {
+
+  options = {
+    services.cdemu = {
+      enable = mkOption {
+        default = false;
```
Comment (lethalman) [inline nixos/modules/services/hardware/cdemu.nix]: Fix indentation here, use spaces.

**Comment by lethalman** [comment]
What's the problem with gnome integration? Be more specific.

File: pkgs/misc/emulators/cdemu/analyzer.nix [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+  version = "3.0.0";
+  pkgName = "image-analyzer";
+  pkgSha256 = "1rb3f7c08dxc02zrwrkfvq7qlzlmm0kd2ah1fhxj6ajiyshi8q4v";
+};
+in callPackage pkg {
+  buildInputs = [ gtk3 libxml2 gnuplot (callPackage ./libmirage.nix {}) makeWrapper ];
+  drvParams = {
+    postFixup = ''
+      wrapProgram $out/bin/image-analyzer \
+        --set XDG_DATA_DIRS "$out/share:${gtk3}/share/gsettings-schemas/${gtk3.name}:$XDG_DATA_DIRS"
```
Comment (bendlas) [inline pkgs/misc/emulators/cdemu/analyzer.nix]: Ad. gsettings schema:

When running the image-analyzer program from command line, it failed with `GLib-GIO-ERROR **: Settings schema 'org.gtk.Settings.FileChooser' is not installed`, while using the .desktop file from the gnome launcher was fine.

I found out, that the environment - var in question was `XDG_DATA_DIRS=/nix/store/j1pfpflc5v970c6rvs2qx9dxzzm7cxrj-gtk+3-3.12.2/share/gsettings-schemas/gtk+3-3.12.2`, so I set up the var like you see in analyzer.nix

For the other packages, It sufficed to forward vars from the build environment and maybe I missed some builtin configuration mechanism. I'm not sure how gnome-shell sets up `XDG_DATA_DIRS` and whether nixos can do the same.

**Comment by bendlas** [comment]
Did the build fail because kernel config exceeded log length?

I will address the other issues shortly.

File: pkgs/misc/emulators/cdemu/analyzer.nix (hunk 3 above)
Comment (lethalman) [inline pkgs/misc/emulators/cdemu/analyzer.nix]: Use `$GSETTINGS_SCHEMAS_PATH`, read other gtk3 derivations, any under desktops/gnome-3. Also add glib to buildInputs.

**Comment by lethalman** [comment]
Ignore travis, staging was recently meged so it's trying to build the world.

**Comment by bendlas** [comment]
Fixed. One issue remains: The file chooser doesn't display the trash icon and there is a warning `Gtk-WARNING **: Could not find the icon 'user-trash-full-symbolic'. The 'hicolor' theme
was not found either, perhaps you need to install it.`

If I add the hicolor theme to buildInputs, the warning goes away, but the trash icon still isn't displayed properly. I wouldn't like to depend on a gnome-icon theme, because this is not a gnome app. Any idea?

Out of curiosity: Why do I need to add glib to buildInputs, when it's in propagatedBuildInputs of gtk3?

**Comment by lethalman** [comment]
If it's propagated no need to add glib yes. If you want an icon by default you must depend on something :) Otherwise leave it as is.

**Comment by bendlas** [comment]
OK, I'm playing around with gnome3.gnome_icon_theme_symbolic, it doesn't seem to pull anything else from gnome, so I might as well use it. I'm a bit confused on how to configure it, though:

- `$XDG_ICON_DIRS` expands to nothing at build time
- If I use `--prefix XDG_DATA_DIRS : "$XDG_DATA_DIRS"`, it finds the icons, however, if I `--set` it, no icons are found, same as if the icon theme wasn't in buildInputs
- If I `--set GDK_PIXBUF_MODULES_FILE "$GDK_PIXBUF_MODULE_FILE"`, it ignores the gnome icon theme and uses the hicolor one (where the trash icon is also missing)

**Comment by bendlas** [comment]
Added a commit to use the icon theme. I'm still not quite sure about it, but it seems to work fine on my gnome3 desktop.

**Comment by lethalman** [comment]
Also add gnome_icon_theme, gdkpixbuf and librsvg.

**Comment by bendlas** [comment]
Done. Is that an implicit dependency of the icon themes?

**Comment by lethalman** [comment]
gnome-icon-theme has the setup-hook to fill XDG_ICON_DIRS. This is going to be moved in hicolor-icon-theme instead, ready in the gnome 3.14 branch. gdkpixbuf has the setup-hook for filling the pixbuf loader, so adding librsvg will add svg loaders.

**Comment by bendlas** [comment]
Oh, I see. Now I can use XDG_ICON_DIRS and don't need to republish all of XDG_DATA_DIRS.
cheers

**Comment by lethalman** [comment]
Built locally thanks.

**Pull request merged**